### PR TITLE
Add Ethereum

### PR DESCRIPTION
**Pull request by paullinator**
This is built on top of PR https://github.com/EdgeApp/edge-currency-accountbased/pull/8 which adds EOS functionality. Merge that PR first.

### CODE REVIEW

**Comment by swansontec** [comment]
This still has the commits from the EOS PR mixed in for some reason. Could you rebase it on top of master for review purposes?

File: src/ethereum/ethEngine.js [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+  decrementNonce: boolean
+}
+
+export class EthereumEngine extends CurrencyEngine {
+  ethereumPlugin: EthereumPlugin
+  otherData: EthereumWalletOtherData
+
+  constructor (
+    currencyPlugin: EthereumPlugin,
+    io_: any,
```
Comment (swansontec) [inline src/ethereum/ethEngine.js]: `EdgeIo` type here.

File: src/ethereum/ethEngine.js [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+      if (walletInfo.keys.keys && walletInfo.keys.keys.ethereumKey) {
+        this.walletInfo.keys.ethereumKey = walletInfo.keys.keys.ethereumKey
+      }
+    }
+    this.currencyPlugin = currencyPlugin
+  }
+
+  async fetchGetEtherscan (server: string, cmd: string) {
+    let apiKey = ''
+    if (global.etherscanApiKey && global.etherscanApiKey.length > 5) {
```
Comment (swansontec) [inline src/ethereum/ethEngine.js]: Note to self: s/global/initOptions/ in the upcoming refactor.

File: src/ethereum/ethEngine.js [hunk 3]
```diff
@@ -0,0 +1,947 @@
...
+    const url = `${server}/api${cmd}${apiKey}`
+    return this.fetchGet(url)
+  }
+
+  async fetchGet (url: string) {
+    const response = await this.io.fetch(url, {
+      method: 'GET'
+    })
+    if (!response.ok) {
+      const cleanUrl = url.replace(global.etherscanApiKey, 'private')
```
Comment (swansontec) [inline src/ethereum/ethEngine.js]: Do we need to sanitize the blockcypher API key here too?

File: src/ethereum/ethEngine.js [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+    this.addToLoop('checkTransactionsInnerLoop', TRANSACTION_POLL_MILLISECONDS)
+    this.addToLoop('checkUnconfirmedTransactionsInnerLoop', UNCONFIRMED_TRANSACTION_POLL_MILLISECONDS)
+    super.startEngine()
+  }
+
+  async killEngine () {
+    // Set status flag to false
+    this.engineOn = false
+    // Clear Inner loops timers
+    for (const timer in this.timers) {
```
Comment (swansontec) [inline src/ethereum/ethEngine.js]: The logic for killing timers seems pretty similar between plugins. Might make sense to put in the base class.

File: src/ethereum/ethEngine.js [hunk 5]
```diff
@@ -0,0 +1,947 @@
...
+    // this.leavePool()
+  }
+
+  async resyncBlockchain (): Promise<void> {
+    await this.killEngine()
+    await this.clearBlockchainCache()
+    await this.startEngine()
+  }
+
+  // synchronous
```
Comment (swansontec) [inline src/ethereum/ethEngine.js]: The comment lies. It's async.

File: src/common/utils.js [hunk 6]
```diff
@@ -12,6 +12,14 @@ function normalizeAddress (address: string) {
   return address.toLowerCase().replace('0x', '')
 }
 
+function addHexPrefix (value: string) {
+  if (value.startsWith('0x')) {
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/common/utils.js]: Updated to use indexOf

File: src/ethereum/ethEngine.js (hunk 3 above)
Comment (paullinator) [reply src/ethereum/ethEngine.js]: Good point. We were never doing that even in the original ethereum plugin.

File: src/ethereum/ethPlugin.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    if (!currencyEngine.otherData.recommendedFee) {
+      currencyEngine.otherData.recommendedFee = '0'
+    }
+
+    const out: EdgeCurrencyEngine = currencyEngine
+    return out
+  }
+
+  async parseUri (uri: string): Promise<EdgeParsedUri> {
+    const networks = { ethereum: true, ether: true }
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/ethereum/ethPlugin.js]: Hmm. are you referring to `parseUriCommon()`. That it does use.

File: src/ethereum/ethEngine.js (hunk 4 above)
Comment (paullinator) [reply src/ethereum/ethEngine.js]: Good catch. moved killEngine to the base class

File: src/ethereum/ethEngine.js (hunk 5 above)
Comment (paullinator) [reply src/ethereum/ethEngine.js]: Removed all these obsolete comments

**Pull request merged**